### PR TITLE
feature/BNGP-3552-legal-agreement-check-and-submit- initial checkin

### PR DESCRIPTION
**Pull request by soorajvdefra**
This PR includes  changes from 3551. https://eaflood.atlassian.net/browse/BNGP-3551 .please do not merge https://github.com/DEFRA/biodiversity-net-gain-service/pull/466

### CODE REVIEW

File: packages/webapp/src/routes/land/check-and-submit.js [hunk 1]
```diff
@@ -34,22 +36,45 @@ const handlers = {
...
+    HabitatWorksStartDate: getDateString(request.yar.get(constants.redisKeys.ENHANCEMENT_WORKS_START_DATE_KEY), 'start date'),
+    HabitatWorksEndDate: getDateString(request.yar.get(constants.redisKeys.LEGAL_AGREEMENT_END_DATE_KEY), 'end date'),
+
     ...geospatialOrLandBoundaryContext(request)
   }
 }
+const getFileNameByType = (files, desiredType) => {
+  const file = files.find(file => file.fileType === desiredType)
+  return file.fileName
+}
```
Comment (pwadmore-ea) [inline packages/webapp/src/routes/land/check-and-submit.js:68]: Use of this function in check-and-submit.html results in display of [object Object] for most upload types. Please perform some refactoring here or elsewhere in this pull request to ensure file names are displayed.

File: packages/webapp/src/views/common/legal-agreement-summary.html [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    },
+    {
+      'text': 'Local land charge search certificate file uploaded',
+      'value': localAndChargeFileName,
+      'valueDataTestId': 'local-land-charge-file-name-value',
+      'valueId': 'add-local-land-charge-file-name-value',
+      'href': routes.CHECK_LOCAL_LAND_CHARGE_FILE,
+      'visuallyHiddenText':  ' local land charge file',
+      'classes': 'govuk-!-margin-bottom-7',
+      'show':showLocalAndLand
```
Comment (pwadmore-ea) [inline packages/webapp/src/views/common/legal-agreement-summary.html:95]: Suggested rename to showLocalLandCharge

File: packages/webapp/src/views/common/legal-agreement-summary.html [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        'valueDataTestId': 'habitat-works-end-date-value',
+        'valueId': 'add-habitat-works-end-date-value',
+        'href': '/land/legal-agreement-end-date',
+        'visuallyHiddenText':  ' legal agreement end date',
+        'classes': 'govuk-!-margin-bottom-7',
+        'show':true
+    },
+    {
+      'text': 'Local land charge search certificate file uploaded',
+      'value': localAndChargeFileName,
```
Comment (pwadmore-ea) [inline packages/webapp/src/views/common/legal-agreement-summary.html:89]: Typo - local**L**AndChargeFileName. This looks like it would need correction throughout the code base.

**Review by pwadmore-ea — CHANGES_REQUESTED** [review]
Please note my comment on the story about disabling the ability to change the legal agreement type temporarily. This is awaiting product owner approval.

File: packages/webapp/src/routes/land/check-and-submit.js (hunk 1 above)
Comment (soorajvdefra) [reply packages/webapp/src/routes/land/check-and-submit.js:68]: this was fixed already on friday

File: packages/webapp/src/views/common/legal-agreement-summary.html (hunk 3 above)
Comment (soorajvdefra) [reply packages/webapp/src/views/common/legal-agreement-summary.html:89]: done

File: packages/webapp/src/views/common/legal-agreement-summary.html (hunk 2 above)
Comment (soorajvdefra) [reply packages/webapp/src/views/common/legal-agreement-summary.html:95]: done

File: packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js [hunk 4]
```diff
@@ -11,54 +11,120 @@ const auth = {
...
     it(`should render the ${url.substring(1)} view`, done => {
       jest.isolateModules(async () => {
         try {
           const session = applicationSession()
-          const getHandler = checkAndSubmit[0].handler
+          let viewResult, contextResult
+          const checkAndSubmitview = require('../../../routes/land/check-and-submit.js')
+          session.set(constants.redisKeys.CONTACT_ID, 'mock contact ID')
+          session.set(constants.redisKeys.APPLICATION_TYPE, 'mock application type')
+          session.set(constants.redisKeys.LEGAL_AGREEMENT_DOCUMENT_TYPE, '759150000')
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js:25]: done

File: packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js [hunk 5]
```diff
@@ -11,54 +11,120 @@ const auth = {
...
+          expect(typeof contextResult.hideClass).toEqual('function')
+          expect(contextResult.hideConsent).toEqual(false)
+          expect(contextResult.routes).not.toBeUndefined()
+          expect(contextResult.changeLandownersHref).toEqual(constants.routes.ADD_LANDOWNERS)
+          expect(contextResult.legalAgreementType).toEqual('Planning obligation (section 106 agreement)')
+          expect(contextResult.legalAgreementFileNames).toEqual('legal-agreement.doc<br>legal-agreement1.pdf')
+          expect(contextResult.responsibleBodies).toEqual('test1,test2')
+          expect(contextResult.HabitatWorksStartDate).toEqual('11 March 2020')
+          expect(contextResult.HabitatWorksEndDate).toEqual('11 March 2024')
+          expect(contextResult.habitatPlanSeperateDocumentYesNo).toEqual('Yes')
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js:81]: done

File: packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js [hunk 6]
```diff
@@ -11,54 +11,120 @@ const auth = {
...
-              redirectArgs = args
+            {
+              location: '800376c7-8652-4906-8848-70a774578dfe/legal-agreement/legal-agreement1.pdf',
+              fileSize: 0.01,
+              fileType: 'application/pdf',
+              id: '2'
             }
-          }
+          ])
+          session.set(constants.redisKeys.HABITAT_PLAN_LOCATION, mockDataPath)
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js:42]: i have removed that line as i am not testing that case

File: packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js [hunk 7]
```diff
@@ -11,54 +11,120 @@ const auth = {
...
+      const h = { redirect: jest.fn() }
+      await handler({ yar: session }, h)
+      expect(h.redirect).toHaveBeenCalledWith(constants.routes.START)
+    })
+    it(`should render the ${url.substring(1)} view with some edge cases`, done => {
+      jest.isolateModules(async () => {
+        try {
+          const session = applicationSession()
+          let viewResult, contextResult
+          const checkAndSubmitview = require('../../../routes/land/check-and-submit.js')
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js:109]: changed to checkAndSubmitView

File: packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js [hunk 8]
```diff
@@ -11,54 +11,120 @@ const auth = {
...
     })
+    it('should redirect to START if APPLICATION_REFERENCE is null', async () => {
+      const session = applicationSession()
+      session.set(constants.redisKeys.APPLICATION_REFERENCE, null)
+      const { handler } = checkAndSubmit.find(route => route.method === 'GET')
+      const h = { redirect: jest.fn() }
+      await handler({ yar: session }, h)
+      expect(h.redirect).toHaveBeenCalledWith(constants.routes.START)
+    })
+    it(`should render the ${url.substring(1)} view with some edge cases`, done => {
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/routes/__tests__/land/check-and-submit.spec.js:104]: changed test description

File: packages/webapp/src/routes/land/check-and-submit.js [hunk 9]
```diff
@@ -34,22 +36,45 @@ const handlers = {
...
+    getLegalAgreementDocumentType(request.yar.get(constants.redisKeys.LEGAL_AGREEMENT_DOCUMENT_TYPE)),
+    legalAgreementFileNames: getLegalAgreementFileNames(applicationDetails.files),
+    responsibleBodies: getResponsibleBodies(request.yar.get(constants.redisKeys.LEGAL_AGREEMENT_RESPONSIBLE_BODIES)),
+    landowners: getLandowners(request.yar.get(constants.redisKeys.LEGAL_AGREEMENT_LANDOWNER_CONSERVATION_CONVENANTS)),
+    habitatPlanSeperateDocumentYesNo: request.yar.get(constants.redisKeys.HABITAT_PLAN_LEGAL_AGREEMENT_DOCUMENT_INCLUDED_YES_NO),
+    getFileNameByType,
+    HabitatPlanFileName: getFileNameByType(applicationDetails.files, 'habitat-plan'),
+    localAndChargeFileName: getFileNameByType(applicationDetails.files, 'local-land-charge'),
+    HabitatWorksStartDate: getDateString(request.yar.get(constants.redisKeys.ENHANCEMENT_WORKS_START_DATE_KEY), 'start date'),
+    HabitatWorksEndDate: getDateString(request.yar.get(constants.redisKeys.LEGAL_AGREEMENT_END_DATE_KEY), 'end date'),
```
[the remark this answers is not in the document]
Comment (soorajvdefra) [reply packages/webapp/src/routes/land/check-and-submit.js]: change link removed fro legal agreement type

File: packages/webapp/src/utils/application-validation.js [hunk 10]
```diff
@@ -2,27 +2,45 @@ import Joi from 'joi'
...
+          is: true,
+          then: Joi.string().allow(null),
+          otherwise: Joi.string().required()
+        }),
+        fileName: Joi.when('optional', {
+          is: true,
+          then: Joi.string().allow(null),
+          otherwise: Joi.string().required()
+        }),
+        optional: Joi.boolean()
```
Comment (pwadmore-ea) [inline packages/webapp/src/utils/application-validation.js:37]: Should this be optional: Joi.boolean().**required()**?

**Review by pwadmore-ea — APPROVED** [review]
Approving based on resolved comments and logged known issues.
